### PR TITLE
Block sgs

### PR DESCRIPTION
**Pull request by mndevec**
Block Gauss Seidel Method is completed.

The method assumes that consecutive *block_size* rows have the same structure. It brings them into shared memory and performs a dot product for the rows that have the same structure.
    
It uses 2 level vectors. Basically, the size of the vector entries can be larger than the available shared memory.
       1) On CPUs: this is not likely to be an issue. Therefore, we increase the size of level-1 vector so that it can hold all entries.    
       2) On GPUs:
             if the required entries is larger than shared memory can hold, we allocated level-2 vector on global memory. This will use memory pools as well. Dot product is called twice first for level-1 vector, then for level-2 vector. Functor tag: BigBlockTeam
          If it fits, we only use level-1 vector. Functor tag is BlockTeam.

Unit tests will practice all cases by reducing the max allowed allocatable shared memory.

### CODE REVIEW

**Comment by mndevec** [comment]
related to https://github.com/kokkos/kokkos-kernels/issues/185

**Comment by srajama1** [comment]
@mndevec : Is this a work in progress or ready to be reviewed and merged ?

**Comment by mndevec** [comment]
@srajama, it is complete. It would be nice if we review and merge it as soon as possible so that I can take care of any problem that might be caught by jenkins-tests.

File: src/sparse/impl/KokkosSparse_gauss_seidel_impl.hpp [hunk 1]
```diff
@@ -220,6 +246,272 @@ class GaussSeidel{
...
+			  Kokkos::parallel_reduce(
+					  Kokkos::ThreadVectorRange(teamMember, l1_val_size),
+					  [&] (nnz_lno_t colind, nnz_scalar_t & valueToUpdate) {
+
+				  valueToUpdate += all_shared_memory[colind] * _adj_vals(current_row_begin + colind);
+
+			  },
+			  product);
+			  //l2 dot product.
+			  //MD: TODO: if thread dot product is implemented, it should be called here again.
```
Comment (srajama1) [inline src/sparse/impl/KokkosSparse_gauss_seidel_impl.hpp]: Can we add a TODO for vector dotproduct as well ?

File: src/sparse/impl/KokkosSparse_gauss_seidel_impl.hpp [hunk 2]
```diff
@@ -220,6 +246,272 @@ class GaussSeidel{
...
+
+
+	  nnz_lno_t *diagonal_positions = (nnz_lno_t *)all_shared_memory;
+	  all_shared_memory =  (nnz_scalar_t *) (((nnz_lno_t *)all_shared_memory) + ((block_size / 8) + 1) * 8);
+
+
+
+	  Kokkos::parallel_for(Kokkos::TeamThreadRange(teamMember, team_row_begin, team_row_end), [&] (const nnz_lno_t& ii) {
+#if 0
+	      Kokkos::single(Kokkos::PerThread(teamMember),[=] () {
```
Comment (srajama1) [inline src/sparse/impl/KokkosSparse_gauss_seidel_impl.hpp]: Another if 0 block

File: perf_test/sparse/KokkosSparse_block_pcg.cpp [hunk 3]
```diff
@@ -0,0 +1,743 @@
...
+  size_t but_r, but_c;
+
+  KokkosKernels::Impl::kk_create_blockcrs_from_blockcrs_formatted_point_crs(
+		  block_size , out_r, out_c,
+		  pf_rm, pf_e, pf_v,
+		  but_r, but_c,
+		  bf_rm, bf_e, bf_v);
+
+  //std::cout << "bnr:" << but_r << " bc:" << but_c<< std::endl;
+#if 0
```
[the remark this answers is not in the document]
Comment (mndevec) [reply perf_test/sparse/KokkosSparse_block_pcg.cpp]: This was for me to debug. I can change all to what you suggested.

File: src/sparse/KokkosSparse_gauss_seidel.hpp [hunk 4]
```diff
@@ -592,6 +623,37 @@ void gauss_seidel_apply_spec(
...
+
+      lno_row_view_t_ row_map,
+      lno_nnz_view_t_ entries,
+      scalar_nnz_view_t_ values,
+      x_scalar_view_t x_lhs_output_vec,
+      y_scalar_view_t y_rhs_input_vec,
+      bool init_zero_x_vector = false,
+      bool update_y_vector = true,
+      int numIter = 1){
+	  handle->get_gs_handle()->set_block_size(block_size);
```
[the remark this answers is not in the document]
Comment (mndevec) [reply src/sparse/KokkosSparse_gauss_seidel.hpp]: Yes, the only difference is the block_size parameter which is 1 vs larger than 1.

File: perf_test/sparse/KokkosSparse_block_pcg.cpp [hunk 5]
```diff
@@ -0,0 +1,743 @@
...
+      delete [] adj;
+      delete [] ew;
+
+      values_view_t kok_x_original = create_x_vector<values_view_t>(((nv /block_size) + 1) * block_size, MAXVAL);
+      for (INDEX_TYPE i = nv; i < ((nv /block_size) + 1) * block_size; ++i){
+    	  kok_x_original(i) = 0;
+      }
+      run_experiment<myExecSpace, crsMat_t>(crsmat, kok_x_original, block_size);
+
+      myExecSpace::finalize();
```
Comment (ndellingwood) [inline perf_test/sparse/KokkosSparse_block_pcg.cpp]: I think calling `finalize()` for a specific execution space is to be deprecated in Kokkos, @crtrott.

File: perf_test/sparse/KokkosSparse_block_pcg.cpp [hunk 6]
```diff
@@ -0,0 +1,743 @@
...
+      delete [] ew;
+
+
+      values_view_t kok_x_original = create_x_vector<values_view_t>(((nv /block_size) + 1) * block_size, MAXVAL);
+      for (INDEX_TYPE i = nv; i < ((nv /block_size) + 1) * block_size; ++i){
+    	  kok_x_original(i) = 0;
+      }
+      run_experiment<myExecSpace, crsMat_t>(crsmat, kok_x_original, block_size);
+
+      myExecSpace::finalize();
```
Comment (ndellingwood) [inline perf_test/sparse/KokkosSparse_block_pcg.cpp]: I think calling finalize() for a specific execution space is to be deprecated in Kokkos, @crtrott.

File: perf_test/sparse/KokkosSparse_block_pcg.cpp [hunk 7]
```diff
@@ -0,0 +1,743 @@
...
+/*
+      for (INDEX_TYPE i = nv; i < ((nv /block_size) + 1) * block_size; ++i){
+    	  kok_x_original(i) = 0;
+      }
+*/
+      run_experiment<myExecSpace, crsMat_t>(crsmat, kok_x_original, block_size);
+
+      //run_experiment<myExecSpace, crsMat_t>(crsmat, kok_x_original);
+
+      myExecSpace::finalize();
```
Comment (ndellingwood) [inline perf_test/sparse/KokkosSparse_block_pcg.cpp]: I think calling finalize() for a specific execution space is to be deprecated in Kokkos, @crtrott.

File: perf_test/sparse/KokkosSparse_block_pcg.cpp [hunk 8]
```diff
@@ -0,0 +1,743 @@
...
+      Kokkos::HostSpace::execution_space::finalize();
```
Comment (ndellingwood) [inline perf_test/sparse/KokkosSparse_block_pcg.cpp]: I think calling finalize() for a specific execution space is to be deprecated in Kokkos (copy/paste previous comment)

**Comment by mndevec** [comment]
@ndellingwood  can you have a look again?

**Comment by mndevec** [comment]
if it looks good, can we merge this branch?

**Comment by crtrott** [comment]
Where is the output from testing? Did you run the tests for this thing on all platforms?

**Comment by crtrott** [comment]
Don't merge yet we are running tests now and need to fix some issues.

**Comment by mndevec** [comment]
White gcc/5.4.0
```
<<Starting on white23>>
Going to test compilers:  gcc/5.4.0
Testing compiler gcc/5.4.0
  Starting job gcc-5.4.0-Serial-release
  Starting job gcc-5.4.0-OpenMP-release
  PASSED gcc-5.4.0-OpenMP-release
  Starting job gcc-5.4.0-OpenMP_Serial-release
  PASSED gcc-5.4.0-Serial-release
  PASSED gcc-5.4.0-OpenMP_Serial-release
#######################################################
PASSED TESTS
#######################################################
gcc-5.4.0-OpenMP-release build_time=282 run_time=240
gcc-5.4.0-OpenMP_Serial-release build_time=409 run_time=932
gcc-5.4.0-Serial-release build_time=243 run_time=745
#######################################################
FAILED TESTS
#######################################################

```

white - cuda/8.0.44
```
<<Starting on white23>>
Running on machine: white
Going to test compilers:  cuda/8.0.44
Testing compiler cuda/8.0.44
  Starting job cuda-8.0.44-Cuda_OpenMP-release
  PASSED cuda-8.0.44-Cuda_OpenMP-release
  Starting job cuda-8.0.44-Cuda_Serial-release
  PASSED cuda-8.0.44-Cuda_Serial-release
#######################################################
PASSED TESTS
#######################################################
cuda-8.0.44-Cuda_OpenMP-release build_time=989 run_time=582
cuda-8.0.44-Cuda_Serial-release build_time=970 run_time=968
#######################################################
FAILED TESTS
#######################################################
```

**Comment by crtrott** [comment]
Let me run a test on Apollo and then we can merge.

**Comment by crtrott** [comment]
Fails with other warnings as errors ;-)

**Comment by crtrott** [comment]
Still fixing more stuff ...

**Comment by crtrott** [comment]
Now this test is failing at runtime (pretty much for all backends):
```
[ RUN      ] serial.graph_graph_color_d2_double_int_int_TestExecSpace
make[2]: *** [test-serial] Floating point exception (core dumped)
```

**Comment by mndevec** [comment]
This branch is still not tested on Jenkins though, right? 
I mean the current failures do not have anything to do with the PR, right?

**Comment by mndevec** [comment]
@crtrott 
The last failure is fixed with this PR: https://github.com/kokkos/kokkos-kernels/pull/201

The base of this PR does not include this fix. If I pull it from develop branch it, the tests pass. Should I pull and push?

**Comment by crtrott** [comment]
So you are saying that floating point exception is fixed on develop? If so then merge develop into the block_sgs branch, push that and then we rerun tests.

**Comment by srajama1** [comment]
@crtrott : Pulling all of develop into this PR will cause a mess if we ever want to revert this. I am ok with pushing this as it is based on Mehmet's comment as this is time sensitive. I would like this to be tested nightly at least couple of days, so I am going to merge later today if you don't have any objections.

**Pull request merged**